storage/buffer.go

```diff
@@ -199,28 +194,95 @@ func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xi
 		}
 
 		unmerged := make([]xio.SegmentReader, 0, len(bucket.encoders))
-		for i := range bucket.encoders {
-			stream := bucket.encoders[i].encoder.Stream()
-			if stream == nil {
-				// TODO(r): log an error and emit a metric, this is pretty bad as this
-				// encoder should have values if "shouldRead" returned true
-				continue
-			}
-
-			ctx.RegisterCloser(func() {
-				// Close the stream and return to pool when read done
-				stream.Close()
-			})
-
+		b.readBucketStreams(ctx, bucket, now, current, func(stream xio.SegmentReader) {
 			unmerged = append(unmerged, stream)
-		}
+		})
 
 		results = append(results, unmerged)
 	})
 
 	return results
 }
 
+func (b *dbBuffer) FetchBlocks(ctx context.Context, starts []time.Time) []FetchBlockResult {
+	var res []FetchBlockResult
+
+	now := b.nowFn()
```

Comment:
Actually `now` is also passed to `readBucketStreams`, so going to leave it as is.